### ISSUE TITLE
Bus Error 

### ISSUE THREAD

**Issue by mahesh-panchal** [marbl/canu#62]
I started a job using Canu, and found out yesterday it crashed with a Bus Error. This is the second job of mine where it has happened.

/var/spool/slurmd/job7088323/slurm_script: line 22: 18990 Bus error               (core dumped) 

Unfortunately because the job is run on a /scratch directory which is cleaned up after exiting, there is no core dump available. The best I can do is give some of the running parameters and where it got to approximately.

The logs have the following (some pieces removed since file names contain research project info):
```
-- Detected Java(TM) Runtime Environment '1.8.0_40' (from 'java').
-- Detected 16 CPUs and 252 gigabytes of memory.
-- Detected Slurm with 'sinfo' binary in /usr/bin/sinfo.
-- Grid engine disabled per useGrid=false option.
--
-- Allowed to run   1 job  concurrently, and use up to  16 compute threads and  252 GB memory for stage 'bogart (unitigger)'.
-- Allowed to run   1 job  concurrently, and use up to  16 compute threads and   48 GB memory for stage 'mhap (overlapper)'.
-- Allowed to run   1 job  concurrently, and use up to  16 compute threads and   48 GB memory for stage 'mhap (overlapper)'.
-- Allowed to run   1 job  concurrently, and use up to  16 compute threads and   48 GB memory for stage 'mhap (overlapper)'.
-- Allowed to run   2 jobs concurrently, and use up to   8 compute threads and   32 GB memory for stage 'read error detection (overlap error adjustment)'.
-- Allowed to run  16 jobs concurrently, and use up to   1 compute thread  and    2 GB memory for stage 'overlap error adjustment'.
-- Allowed to run   2 jobs concurrently, and use up to   8 compute threads and   64 GB memory for stage 'utgcns (consensus'.
-- Allowed to run  16 jobs concurrently, and use up to   1 compute thread  and    8 GB memory for stage 'overlap store sequential building'.
-- Allowed to run  16 jobs concurrently, and use up to   1 compute thread  and    4 GB memory for stage 'overlap store parallel bucketizer'.
-- Allowed to run  16 jobs concurrently, and use up to   1 compute thread  and   15 GB memory for stage 'overlap store parallel sorting'.
-- Allowed to run  16 jobs concurrently, and use up to   1 compute thread  and    8 GB memory for stage 'overlapper'.
-- Allowed to run   2 jobs concurrently, and use up to   8 compute threads and   16 GB memory for stage 'overlapper'.
-- Allowed to run   2 jobs concurrently, and use up to   8 compute threads and   16 GB memory for stage 'overlapper'.
-- Allowed to run   1 job  concurrently, and use up to  16 compute threads and  252 GB memory for stage 'meryl (k-mer counting)'.
-- Allowed to run   2 jobs concurrently, and use up to   8 compute threads and   16 GB memory for stage 'falcon_sense (read correction)'.
--
-- This is canu parallel iteration #1, out of a maximum of 2 attempts.
--
-- Final error rates before starting pipeline:
--   
--   genomeSize          -- 2000000000
--   errorRate           -- 0.6
--   
--   corOvlErrorRate     -- 1.8
--   obtOvlErrorRate     -- 1.8
--   utgOvlErrorRate     -- 1.8
--   
--   obtErrorRate        -- 1.8
--   
--   utgGraphErrorRate   -- 1.2
--   utgBubbleErrorRate  -- 1.5
--   utgMergeErrorRate   -- 0.9
--   utgRepeatErrorRate  -- 1.2
--   
--   corErrorRate        -- 6
--   cnsErrorRate        -- 1.5
--
--
-- BEGIN CORRECTION
--
--
-- GATEKEEPER (correction)
--
----------------------------------------
-- Starting command on Mon Feb  8 17:19:34 2016 with 3439.6 GB free disk space
...
-- Finished on Mon Feb  8 17:38:04 2016 (1110 seconds) with 3431.6 GB free disk space
...
-- MERYL (correction)
-- Meryl attempt 1 begins.
----------------------------------------
-- Starting concurrent execution on Mon Feb  8 17:43:08 2016 with 3431.4 GB free disk space (1 processes; 1 concurrently)
...
-- Finished on Mon Feb  8 18:24:58 2016 (2510 seconds) with 3422.4 GB free disk space
----------------------------------------
-- Meryl finished successfully.
----------------------------------------
-- Starting command on Mon Feb  8 18:25:15 2016 with 3422.4 GB free disk space
... 
-- Finished on Mon Feb  8 18:25:17 2016 (2 seconds) with 3422.4 GB free disk space
----------------------------------------
--
-- Found 27948955677 16-mers; 2046586086 distinct and 196497036 unique.  Largest count 13055842.
--
-- OVERLAPPER (mhap) (correction)
--
-- Given 48 GB, can fit 72000 reads per block.
-- For 125 blocks, set stride to 31 blocks.
-- Logging partitioning to '/scratch/7088323/canu_assembly/correction/1-overlapper/partitioning.log'.
-- Computed seed length 500 from desired output coverage 40 and genome size 2000000000
-- Configured 124 mhap precompute jobs.
-- Configured 307 mhap overlap jobs.
-- mhap precompute attempt 0 begins with 0 finished, and 124 to compute.
----------------------------------------
-- Starting concurrent execution on Mon Feb  8 18:27:41 2016 with 3431.3 GB free disk space (124 processes; 1 concurrently)
...
-- Finished on Tue Feb  9 01:08:21 2016 (24040 seconds) with 3292.8 GB free disk space
----------------------------------------
-- All 124 mhap precompute jobs finished successfully.
-- mhap attempt 0 begins with 0 finished, and 307 to compute.
----------------------------------------
-- Starting concurrent execution on Tue Feb  9 01:08:21 2016 with 3292.8 GB free disk space (307 processes; 1 concurrently)
...
   /scratch/7088323/canu_assembly/correction/1-overlapper/mhap.sh 307 > /scratch/7088323/canu_assembly/correction/1-overlapper/mhap.000307.out 2>&1
/var/spool/slurmd/job7088323/slurm_script: line 22: 18990 Bus error               (core dumped) ...
```

So as far as I can tell, the error occurred just after the last MHAP attempt finished.

**Comment by skoren** [comment]
I've usually seen bus errors when the binary changed while the code was running. What output, if any, is in mhap.000307.out? Canu also shouldn't have stopped on that error, it would have retried the failed job and given up after consistent failure. What was the Canu message after the error and what was your Canu command?

Lastly, the error rate shouldn't be set manually and certainly not to 60% (0.6). It is meant to be the error rate in the corrected reads which is typically 2-3% (0.02) for PacBio data and 4-5% (0.04) in Nanopore data. This shouldn't have caused the bus error but would cause other issues downstream in your compute.

**Comment by mahesh-panchal** [comment]
Hi Sergey,

Any files that were on /scratch no longer exist so I cannot see what the 
output is. The executing script copies the results off /scratch only on 
completion.

The binary shouldn't have changed while the code was running. Nothing 
was attempted with compilation since starting the runs.

The last lines of the log are:
     /scratch/7088323/canu_assembly/correction/1-overlapper/mhap.sh 305 
 > 
/scratch/7088323/canu_assembly/correction/1-overlapper/mhap.000305.out 2>&1
     /scratch/7088323/canu_assembly/correction/1-overlapper/mhap.sh 306 
 > 
/scratch/7088323/canu_assembly/correction/1-overlapper/mhap.000306.out 2>&1
     /scratch/7088323/canu_assembly/correction/1-overlapper/mhap.sh 307 
 > 
/scratch/7088323/canu_assembly/correction/1-overlapper/mhap.000307.out 2>&1
/var/spool/slurmd/job7088323/slurm_script: line 22: 18990 Bus error 
           (core dumped) /proj/b2014264/src/canu/Linux-amd64/bin/canu -p 
Organism_pacbio_canu -d $SNIC_TMP/canu_assembly -genomeSize=2g 
-errorRate=0.6 -useGrid=false -pacbio-raw $PACBIOFILES
(END)

Thank you for the error rate information. I missed that.

That's all the information I have about as far as it got.
Thank you for your time.

Regards,
Mahesh.

**Comment by mahesh-panchal** [comment]
One more thing that could explain the exit is the executing bash script 
has the command

set -e

I'll try removing that and rerunning.

**Comment by HenrivdGeest** [comment]
Just to add something for future readers, we experienced a lot of bus errors on binaries. Not only for Canu. It has to do with our NFS infrastructure, which we haven't solved up-to date. We just changed to another storage device which does not show that error. 
The error only happened when  a binary was running for quite some time., and for sure we did not change the binary in the mean time. Its just our NFS.
